rslib/src/tags.rs

```diff
@@ -2,16 +2,80 @@
 // License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
 
 use crate::{
+    backend_proto::{Tag as TagProto, TagTreeNode},
     collection::Collection,
     err::{AnkiError, Result},
     notes::{NoteID, TransformNoteOutput},
-    text::to_re,
-    {text::normalize_to_nfc, types::Usn},
+    text::{normalize_to_nfc, to_re},
+    types::Usn,
 };
+
 use regex::{NoExpand, Regex, Replacer};
-use std::{borrow::Cow, collections::HashSet};
+use std::cmp::Ordering;
+use std::{
+    borrow::Cow,
+    collections::{HashMap, HashSet},
+    iter::Peekable,
+};
 use unicase::UniCase;
 
+#[derive(Debug, Clone)]
+pub struct Tag {
+    pub name: String,
+    pub usn: Usn,
+    pub collapsed: bool,
+}
+
+impl Ord for Tag {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.name.cmp(&other.name)
+    }
+}
+
+impl PartialOrd for Tag {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for Tag {
+    fn eq(&self, other: &Self) -> bool {
+        self.name == other.name
+    }
+}
+
+impl Eq for Tag {}
+
+impl Default for Tag {
+    fn default() -> Self {
+        Tag {
+            name: "".to_string(),
+            usn: Usn(-1),
```

Comment:
The server should never have a usn of -1, and it's a bit dangerous to be providing a default value. Is this default implementation just to make creating tags easier? If so, an alternative is to create a new() function that takes required functions and fills out other defaults, eg

```rust
impl Tag {
  fn new(name: String, usn: Usn) → Self {
    Tag { name: name.into(), usn, collapsed: false }
  }
}
```

The Into part is not required, it just makes it easier to pass in string